src/hclust.jl

```diff
@@ -531,6 +531,108 @@ function rorder!(hmer::HclustMerges)
     return hmer
 end
 
+
+"""
+    orderleaves_barjoseph!(hmer::HclustMerges, dm::AbstractMatrix)
+
+Given a hierarchical cluster and the distance matrix used to generate it,
+use fast algorithm to determine optimal leaf order minimizing the distance
+between adjacent leaves. This is done using a heuristic where,
+when combining multi-leaf sub branches, only the outermost leaves are
+compared (a maximum of 4 comparisons per intersection).
+
+Sub branches are flipped if necessary to minimize the distance between adjacent
+nodes, and then the combined branches are treated as a block for future
+comparisons.
+
+Based on:
+[Bar-Joseph et. al. "Fast optimal leaf ordering for hierarchical clustering." _Bioinformatics_. (2001)](https://doi.org/10.1093/bioinformatics/17.suppl_1.S22)
+"""
+function orderleaves_barjoseph!(hmer::HclustMerges, dm::AbstractMatrix)
+    order = invperm(hclust_perm(hmer))
+    node_ranges = Tuple{Int,Int}[] # ranges of order array indices occupied by the leaves of each node
+
+    for v in 1:nnodes(hmer)-1
+        vl, vr = hmer.mleft[v], hmer.mright[v]
+
+        (uidx, midx) = node_range(vl, order, node_ranges)
+        (kidx, widx) = node_range(vr, order, node_ranges)
+
+        (u, m) = (order[uidx], order[midx])
+        (k, w) = (order[kidx], order[widx])
+
+        if vl < 0 && vr < 0
+            # Nothing needs to be done
+        elseif vl < 0
+            # check if flipping would reduce distance
+            if dm[m,k] > dm[m,w]
+                reverse!(order, uidx, midx)
+                rotate_merges!(hmer, vr)
+            end
+        elseif vr < 0
+            if dm[k,m] > dm[k,u]
+                reverse!(order, uidx, midx)
+                rotate_merges!(hmer, vl)
+            end
+        elseif vl > 0 && vr > 0
+            # For 2 multi-leaf branches, determine if one or two flips is required
+            # 1 = do not flip
+            # 2 = flip left
+            # 3 = flip right
+            # 4 = flip both
+            flp = argmin((dm[m,k], dm[u,k], dm[m,w], dm[u,w]))
+            if flp == 2 || flp == 4
+                reverse!(order, uidx, midx)
+                rotate_merges!(hmer, vl)
+            end
+            if flp == 3 || flp == 4
+                reverse!(order, kidx, widx)
+                rotate_merges!(hmer, vr)
+            end
+        else
+            error("invalid 'merge' order in Hclust: ($vl, $vr) ")
```

Comment:
```suggestion
            error("invalid 'merge' order in Hclust: ($vl, $vr)")
```
Would this be triggered only when `vl == vr == 0`?

Reply:
Yes, I think that's right. I think when I was originally writing this code, I didn't cover the case where `vl` was positive and `vr` was negative because it seemed like that never happened in merges. I added this `else` to make sure that never happened. Would you prefer I remove it?

The only harm I think is that it hurts codecov percentage :-/

Reply:
I think it's fine to keep this `error()`. I personally don't know a better alternative.
Maybe just add
```julia
else # vl == vr == 0
```
to help understanding the code.